### PR TITLE
added encode_www_form to make sure the path helpers are properly escaping dynamic segments

### PR DESCRIPTION
**Pull request by foxnewsnetwork**
This is in reference to:

https://github.com/phoenixframework/phoenix/issues/1286

@josevalim suggested to put the `encode_www_form` inside `to_param`, however, because we already use `to_param` to properly encode querystrings, doing so resulted in a double encoding. The next best place was to place the `encode_www_form` directly inside the `expand_segments` function.

Admittedly, I am a little surprised all the tests passed despite the fact I didn't also add `encode_www_form` to L238. I can spend a little more time trying to break into corner cases, but here's the working pr for now.

### CODE REVIEW

**Comment by josevalim** [comment]
Ah, good catch. to_param is called from other places and i missed that!

**Comment by josevalim** [comment]
@foxnewsnetwork we need to change line 238 too. Maybe by using `Enum.map_join` and passing encode as the mapper function. That clause is called when you have `/foo/*path` segments. I hope this info helps you write another failing test. :)

**Comment by foxnewsnetwork** [comment]
Got it, thank you. Will get on it.

**Comment by foxnewsnetwork** [comment]
okay, as per your instructions, I replaced `Enum.join` with `Enum.map_join` and added a bunch more tests. 

I was, at first, concerned with the fact I didn't need to change the code at

https://github.com/phoenixframework/phoenix/blob/v1.0.3/lib/phoenix/router/helpers.ex#L238

to also include `URI.encode_www_form`, but after tinkering around a bit more with the code, I realized that part would only require changing if someone decides to use odd characters in the non-dynamic parts of their route in the router like so:

```elixir
get "/dog+=3ad/:id", DogController, :show
```
As this use-case seems rather silly, I opted to not alter the code at L238 nor to write tests for this.

File: test/phoenix/router/helpers_test.exs [hunk 1]
```diff
@@ -152,6 +152,22 @@ defmodule Phoenix.Router.HelpersTest do
     end
   end
 
+  test "top-level named routes with complex ids" do
+    assert Helpers.post_path(__MODULE__, :show, "8==D--+") ==
```
Comment (josevalim) [inline test/phoenix/router/helpers_test.exs]: The choice of characters `8==D` do not really belong to a technical codebase. If you haven't noticed it, it is OK because I haven't either the first time I reviewed it. Can we pick something else? Dropping the number 8 would be fine.

**Comment by foxnewsnetwork** [comment]
Acknowledged, it is done. 

On an unrelated note, I noticed the following test would fail on running the tests after restarting my computer:

```bash
1) test starts watching and writes to stdio (Phoenix.Endpoint.WatcherTest)
     test/phoenix/endpoint/watcher_test.exs:7
     No message matching {:DOWN, ^ref, :process, ^pid, :normal} after 100ms. The process mailbox is empty.
     stacktrace:
       (ex_unit) lib/ex_unit/capture_io.ex:99: ExUnit.CaptureIO.do_capture_io/3
       test/phoenix/endpoint/watcher_test.exs:8
```
I'm pretty sure no amount of strings comparing "8"s to "D"s on my part could be causing this... can it be 100ms is too short on a fresh ram computer?

**Comment by josevalim** [comment]
@foxnewsnetwork it is a race condition somewhere. :(

**Comment by foxnewsnetwork** [comment]
@josevalim I can't reproduce this fail test on my machine any more, regardless of restarts or whatever. I will try doing it on travis

**Comment by josevalim** [comment]
Merged manually, thank you!  :heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by foxnewsnetwork** [comment]
np (I'm sorry I occasionally have the mind of a 12 year old lol).

I have no idea what could be causing the race condition, but I've managed to reproduce it after rebuilding a couple of times on travis:

https://travis-ci.org/foxnewsnetwork/phoenix/builds/85776438#L335

I have no idea what could be causing this. But it seems to just be capture_io related, so could it be that this bug isn't even in phoenix?

**Comment by josevalim** [comment]
Elixir v1.1 fixed a race condition on capture_io, so it may as well be fixed in there. :)
